primitives/src/signatures/bls_over_bn254.rs

```diff
@@ -0,0 +1,398 @@
+    pub fn sign_key_internal(&self) -> &ScalarField {
+        &self.sk.0
+    }
+
+    /// Signature function
+    pub fn sign<B: AsRef<[u8]>>(&self, msg: &[u8], csid: B) -> Signature {
+        let msg_input = [msg, csid.as_ref()].concat();
+        let hash_value: G1Projective = hash_to_curve::<Keccak256>(&msg_input);
+        let sigma = hash_value * self.sk.0;
+        Signature { sigma }
+    }
+}
+
+impl SignKey {
+    fn generate<R: Rng>(prng: &mut R) -> SignKey {
+        SignKey(ScalarField::rand(prng))
+    }
+}
+
+impl From<&SignKey> for VerKey {
+    fn from(sk: &SignKey) -> Self {
+        VerKey(G2Projective::generator() * sk.0)
+    }
+}
+
+impl VerKey {
+    /// Get the internal of verifying key, namely a curve Point
+    pub fn internal(&self) -> G2Projective {
+        self.0
+    }
+
+    /// Signature verification function
+    pub fn verify<B: AsRef<[u8]>>(
+        &self,
+        msg: &[u8],
+        sig: &Signature,
+        csid: B,
+    ) -> Result<(), PrimitivesError> {
+        let msg_input = [msg, csid.as_ref()].concat();
+        let group_elem = hash_to_curve::<Keccak256>(&msg_input);
+        let g2 = G2Projective::generator();
+
+        let is_sig_valid = Bn254::pairing(sig.sigma, g2) == Bn254::pairing(group_elem, self.0);
+        match is_sig_valid {
+            true => Ok(()),
+            false => Err(VerificationError("Pairing check failed".to_string())),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    // These tests are adapted from schnorr.rs
+    use crate::{
+        constants::CS_ID_BLS_BN254,
+        signatures::{
+            bls_over_bn254::{
+                BLSOverBN254CurveSignatureScheme, KeyPair, SignKey, Signature, VerKey,
+            },
+            tests::{failed_verification, sign_and_verify},
+        },
+    };
+    use ark_ff::vec;
+    use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+    use jf_utils::Vec;
+
+    #[test]
+    fn test_bls_signature_internals() {
+        let mut rng = jf_utils::test_rng();
+        let key_pair1 = KeyPair::generate(&mut rng);
+        let key_pair2 = KeyPair::generate(&mut rng);
+        let key_pair3 = KeyPair::generate(&mut rng);
+        let pk_bad: VerKey = KeyPair::generate(&mut rng).ver_key();
+        let key_pairs = [key_pair1, key_pair2, key_pair3];
+
+        let mut msg = vec![];
+        for i in 0..10 {
+            for key_pair in &key_pairs {
+                assert_eq!(key_pair.vk, VerKey::from(&key_pair.sk));
+                let sig = key_pair.sign(&msg, CS_ID_BLS_BN254);
+                let pk = key_pair.ver_key_ref();
+                assert!(pk.verify(&msg, &sig, CS_ID_BLS_BN254).is_ok());
+                // wrong public key
+                assert!(pk_bad.verify(&msg, &sig, CS_ID_BLS_BN254).is_err());
+                // wrong message
+                msg.push(i as u8);
+                assert!(pk.verify(&msg, &sig, CS_ID_BLS_BN254).is_err());
+            }
+        }
+    }
```

Comment:
Shall we additionally test 2 messages m1, m2=(m1||x) whose lengths are larger than a base field length? E.g., |m1| = 35B, |x| = 1B

Reply:
Yes it is a good idea, but why do we need two messages? Is it not enough to test with m1?

Reply:
E.g:

```rust
// Test for long messages
let size=35; // Bigger than 32 which is the number of bytes needed to encode a field element
let key_pair = KeyPair::generate(&mut rng);
let mut msg = [33u8;35];
let sig = key_pair.sign(&msg, CS_ID_BLS_BN254);
let pk = key_pair.ver_key_ref();
assert!(pk.verify(&msg, &sig, CS_ID_BLS_BN254).is_ok());
```

Reply:
To check that a signature for m1 won't verify for m2, or vice versa.

Reply:
The above code looks good. Maybe setup a `msg2 = [33u8; 36]` and add one more line at the bottom:
`assert!(pk.verify(&msg2, &sig, CS_ID_BLS_BN254).is_err());`

Reply:
Ok, done in a07c2293.